tests/infra/module_tester.py

```diff
@@ -0,0 +1,141 @@
+
+    It supports testing inference and training modes.
+    """
+
+    def __init__(
+        self,
+        test_type: TestType = TestType.INFERENCE,
+        comparison_metric: ComparisonMetric = ComparisonMetric.PCC,
+    ) -> None:
+        self._test_type = test_type
+        self._comparison_metric = comparison_metric
+
+    def __call__(self, module: TestModule) -> bool:
+        """
+        The only public method providing testing hook.
+
+        Call tester with passed test module and it will run tests on it.
+        """
+        return self._test(module)
+
+    def _test(self, module: TestModule) -> bool:
+        if self._test_type == TestType.INFERENCE:
+            return self._test_inference(module)
+        else:
+            return self._test_training(module)
+
+    def _test_inference(self, module: TestModule) -> bool:
+        tt_res = DeviceRunner.run_on_tt_device(module)
+        cpu_res = DeviceRunner.run_on_cpu(module)
+        return self._compare(tt_res, cpu_res)
+
+    def _test_training(self, module: TestModule) -> bool:
+        raise NotImplementedError("Support for training not implemented")
+
+    def _compare(
+        self, device_out: jax.Array, golden_out: jax.Array, assert_on_fail: bool = False
+    ) -> bool:
+        device_output, golden_output = DeviceRunner.put_on_cpu(device_out, golden_out)
+        device_output, golden_output = self._match_data_types(
+            device_output, golden_output
+        )
+
+        if self._comparison_metric == ComparisonMetric.EQUAL:
+            comp = compare_equal(device_output, golden_output)
+        elif self._comparison_metric == ComparisonMetric.PCC:
+            comp = compare_pcc(device_output, golden_output)
+        elif self._comparison_metric == ComparisonMetric.ATOL:
+            comp = compare_atol(device_output, golden_output)
+        elif self._comparison_metric == ComparisonMetric.ALLCLOSE:
+            comp = compare_allclose(device_output, golden_output)
+
+        if assert_on_fail:
+            assert comp, f"{self._comparison_metric.value} comparison failed!"
+
+        return comp
+
+    def _match_data_types(self, *tensors: jax.Array) -> Sequence[jax.Array]:
+        """
+        Casts all tensors to float32 if not already in that format.
+
+        Tensors need to be in same data format in order to compare them.
+        """
+        return [
+            tensor.astype("float32") if tensor.dtype.str != "float32" else tensor
+            for tensor in tensors
+        ]
+
+
+def _test(
+    f: Union[Callable, TestModel],
+    inputs: Sequence[jax.Array],
+    comparison_metric: ComparisonMetric,
+) -> bool:
+    """Helper 'protected' method, don't use, use provided public methods below instead."""
+    tester = ModuleTester(comparison_metric=comparison_metric)
+    module = (
+        f.as_test_module(inputs)
+        if isinstance(f, TestModel)
+        else TestModule(f, args=inputs)
+    )
+    return tester(module)
+
+
+def test(
```

Comment:
Prefixing public functions with `test` will cause trouble with `pytest -v tests`. If you have a better name for this and function below, please suggest @mrakitaTT @ajakovljevicTT .

Reply:
I also need you opinion on https://github.com/tenstorrent/tt-xla/pull/101/files#r1869054172.

Reply:
Couple of thoughts here:
- We should not have a `.sh` file to run tests which we have to maintain, it is better if we can use plain `pytest` call to run tests. Therefore we should avoid using `test` prefix, and use something like `run_test`.
- It would be better to not even have global functions like this, and instead for this method to be inside the ModuleTester or ModelTester. I really like how tt-torch [ModelTester](https://github.com/tenstorrent/tt-torch/blob/main/tests/utils.py#L15) has separate functions to load model, compile model, run the model in eval/train modes, verify results, etc. Tester for ops does not need to have any of this, which is why I think we are shooting ourselves in the foot here by trying to make something universal to work for both ops and models testing. Let's talk more about this today.
- In the future tt-xla will be used for other frameworks too, so we should create a bit different structure for tests (not as part of this PR though):
```
tests/
  infra/
    (infra files)
  jax/
    ops/
    graphs/
    models/
    test_data_types.py
    ...
```